File: src/main/java/com/avilapps/pdf_services/domain/services/DocumentService.java
package com.avilapps.pdf_services.domain.services;

import com.avilapps.pdf_services.common.exceptions.NonImplementedException;
import com.avilapps.pdf_services.domain.model.ContentFile;
import com.avilapps.pdf_services.domain.model.Document;

public interface DocumentService {

    default Document foliate(Document document) {
        throw new NonImplementedException();
    }

    default String upload(ContentFile contentFile) {
        throw new NonImplementedException();
    }
}

File: src/main/kotlin/com/avilapps/pdf_services/domain/service/DocumentService.kt
package com.avilapps.pdf_services.domain.service

import com.avilapps.pdf_services.domain.repository.DocumentRepository
import org.springframework.stereotype.Service

@Service
class DocumentService(private val documentRepository: DocumentRepository, private val foliationService: FoliationService) {

    fun upload(destinationPath: String, content: ByteArray): String {
        documentRepository.upload(destinationPath, content)
        return documentRepository.getPublicUrl(destinationPath).toString()
    }

    fun foliate(originPath: String, initialFolio: Int, destinationPath: String): String {
        val originalContent = documentRepository.download(originPath)
        val foliatedContent = foliationService.foliate(originPath, originalContent, initialFolio)
        documentRepository.upload(destinationPath, foliatedContent)
        return documentRepository.getPublicUrl(destinationPath).toString()
    }
}